gamemode/modules/crafting/sv_crafting.lua

```diff
@@ -0,0 +1,136 @@
+RVR.Crafting = RVR.Crafting or {}
+local cft = RVR.Crafting
+
+util.AddNetworkString( "RVR_Crafting_AttemptCraft" )
+util.AddNetworkString( "RVR_Crafting_CraftResponse" )
+util.AddNetworkString( "RVR_Crafting_CloseCraftingMenu" )
+util.AddNetworkString( "RVR_Crafting_OpenCraftingMenu" )
+util.AddNetworkString( "RVR_Crafting_Grab" )
+
+local function craftFail( ply )
+    net.Start( "RVR_Crafting_CraftResponse" )
+        net.WriteInt( cft.STATE_WAITING, 4 )
+    net.Send( ply )
+end
+
+net.Receive( "RVR_Crafting_Grab", function( len, ply )
+    local ent = ply.RVR_CraftingEnt or ply
+
+    if not ent.RVR_Crafting then return end
+
+    local state = cft.STATE_WAITING
+
+    local output = ent.RVR_Crafting.output
+    if output then
+        if CurTime() >= output.timeStart + output.recipe.timeToCraft then
+            local recipe = output.recipe
+            local itemInstance = RVR.Items.getItemInstance( recipe.item )
+
+            if RVR.Inventory.canFitItem( ply.RVR_Inventory, itemInstance, recipe.count ) then
+                RVR.Inventory.attemptPickupItem( ply, itemInstance, recipe.count )
+                ent.RVR_Crafting.output = nil
+            else
+                state = cft.STATE_CRAFTED
+            end
+        else
+            state = cft.STATE_CRAFTING
+        end
+    end
+
+    net.Start( "RVR_Crafting_CraftResponse" )
+        net.WriteInt( state, 4 )
+    net.Send( ply )
+end )
+
+net.Receive( "RVR_Crafting_AttemptCraft", function( len, ply )
+    local categoryID = net.ReadInt( 8 )
+    local recipeID = net.ReadInt( 8 )
+
+    local category = RVR.Crafting.Recipes[categoryID]
+    if not category then return craftFail( ply ) end
+
+    local recipe = category.recipes[recipeID]
+    if not recipe then return craftFail( ply ) end
+
+    local craftingEnt = ply.RVR_CraftingEnt or ply
+
+    if not cft.craft( ply, craftingEnt, recipe ) then craftFail( ply ) end
+end )
+
+function cft.craft( ply, ent, recipe )
+    if not ent.RVR_Crafting then return end
+
+    if ent.RVR_Crafting.output then return end
+
+    local tier = ent.RVR_Crafting.tier
+    if tier < recipe.tier then return end
+
+    local category = RVR.Crafting.Recipes[recipe.categoryID]
+    if ent.RVR_Crafting.type ~= category.crafterType then return end
+
+    local success = RVR.Inventory.tryTakeItems( ply, recipe.itemsStruct )
+    if not success then return end
+
+    ent.RVR_Crafting.output = {
+        recipe = recipe,
+        timeStart = CurTime(),
+    }
+
+    net.Start( "RVR_Crafting_CraftResponse" )
+        net.WriteInt( cft.STATE_CRAFTING, 4 )
+    net.Send( ply )
+
+    return true
```

Comment:
if this returns  a boolean, the other returns should return false instead of nil